frontend/08 Reselect/README.md

```diff
@@ -1,24 +1,214 @@
-# 08 Reselect
-
-In this sample we are going to testing selectors.
-
-We will start from sample _07 Code Coverage_.
-
-Summary steps:
- - .
-
-# Steps to build it
-
-- `npm install` to install previous sample packages:
-
-```bash
-npm install
-
-```
-
-# About Lemoncode
-
-We are a team of long-term experienced freelance developers, established as a group in 2010.
-We specialize in Front End technologies and .NET. [Click here](http://lemoncode.net/services/en/#en-home) to get more info about us.
-
-For the LATAM/Spanish audience we are running an Online Front End Master degree, more info: http://lemoncode.net/master-frontend
+# 08 Reselect
+
+In this sample we are going to testing selectors.
+
+We will start from sample _07 Code Coverage_.
+
+Summary steps:
+ - Install `reselect`.
+ - Create memoized selector for `members`.
+
+# Steps to build it
+
+- `npm install` to install previous sample packages:
+
+```bash
+npm install
+```
+
+Let's start adding `reselect` to our `package.json`:
+
+```bash
+npm install -S reselect
+```
+
+TypeScript definitions are included inside `reselect` so we don't need to install extra package from DefinitelyTyped (`@types`).
+
+First of all we'll identify what selectors does really need `reselect`. Let's have a look at our containers:
+
+- `pages/login/pageContainer.tsx` is using selectors directly. We could apply `reselect` but we don't gain anything from that since `mapStateToProps` is getting directly pointers to those objects from state.
+- `pages/members/pageContainer.tsx` on the other hand is using a mapper to transform members list on every state change. It is a good candidate for a memoized selector because if there are multiple mounted containers requesting same mapped list the list could be computed once and be served by as many containers as needed.
+
+First of all let's create a dummy function that create an empty state object called `mockInitialState.ts`:
+
+#### `test/mockInitialState.ts`
+
+```ts
+import { State } from '../src/pages/reducers';
+import { createEmptyLoginEntity, createEmptyLoginFormErrors } from '../src/pages/login/viewModel';
+
+export const mockInitialState = (): State => ({

```

Comment:
I think this file is not necessary. It could be a bit confused.

When we create unit tests, we try to keep defined all variables in `Arrange` section to know with the naked eye all necessary stuff to run that test.

Is it not enough if we declare something like:

```
it('should return members from state', () => {
      // Arrange
      const state = {
        members: [
          {
            id: 1,
            login: 'login member 1',
            avatar_url: 'avatar member 1',
          },
        ]
      } as State;

      // Act
      const result = getMembers(state);

      // Assert
      expect(result).toEqual(state.members);
    });
```